File: leetcode/105_Construct_Binary_Tree_from_Preorder_and_Inorder_Traversal.py
from typing import List, Optional
import unittest

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

    def __eq__(self, obj):
        return isinstance(obj, TreeNode) and self.val == obj.val and self.left == obj.left and self.right == obj.right

    def __repr__(self):
        return f"({self.val} {self.left} {self.right})"

class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        def _rec(in_i: int, in_j: int) -> Optional[TreeNode]:
            if  in_i > in_j:
                return None
            nonlocal current_val_idx
            val = preorder[current_val_idx]
            inorder_val_idx = inorder_idx[val]
            current_val_idx += 1
            return TreeNode(
                val,
                _rec(in_i, inorder_val_idx - 1),
                _rec(inorder_val_idx + 1, in_j)
            )
        inorder_idx = {value: idx for idx, value in enumerate(inorder)}
        n = len(preorder)
        current_val_idx = 0
        return _rec(0, n - 1)

class BuildTreeTest(unittest.TestCase):

    def test_1(self) -> None:
        self.assertEqual(
            Solution().buildTree([3,9,20,15,7], [9,3,15,20,7]),
            TreeNode(3, TreeNode(9), TreeNode(20, TreeNode(15), TreeNode(7)))
            ) 

    def test_2(self) -> None:
        self.assertEqual(
            Solution().buildTree([-1], [-1]),
            TreeNode(-1)
            ) 
    def test_3(self) -> None:
        self.assertEqual(
            Solution().buildTree([1, 2], [2, 1]),
            TreeNode(1, TreeNode(2))
            ) 

File: leetcode/105_Construct_Binary_Tree_from_Preorder_and_Inorder_Traversal.rs
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None
        }
    }
}

struct Solution {}

use std::rc::Rc;
use std::cell::RefCell;
use std::collections::HashMap;

impl Solution {
    pub fn build_tree(preorder: Vec<i32>, inorder: Vec<i32>) -> Option<Rc<RefCell<TreeNode>>> {
        pub fn _rec(preorder: &Vec<i32>, inorder: &Vec<i32>, first: isize, last: isize, inorder_idx_dic: &HashMap<i32, isize>, preorder_idx: &mut usize) -> Option<Rc<RefCell<TreeNode>>> {
            if first > last {
                None
            } else {
                let val = preorder[*preorder_idx];
                let inorder_idx = inorder_idx_dic[&val];
                *preorder_idx += 1;
                Some(Rc::new(RefCell::new(
                    TreeNode {
                        val,
                        left: _rec(preorder, inorder, first, inorder_idx  - 1, inorder_idx_dic, preorder_idx),
                        right: _rec(preorder, inorder, inorder_idx  + 1, last, inorder_idx_dic, preorder_idx)
                    }
                )))
            }
        }
        let mut inorder_idx_dic = HashMap::new();
        for i in 0..inorder.len() {
            inorder_idx_dic.insert(inorder[i], i as isize);
        }
        let mut preorder_idx = 0;
        _rec(&preorder, &inorder, 0, inorder.len() as isize - 1, &inorder_idx_dic, &mut preorder_idx)
    }
}

#[cfg(test)]
mod build_tree_test {
    use super::*;

    fn node(val: i32, left: Option<Rc<RefCell<TreeNode>>>, right: Option<Rc<RefCell<TreeNode>>>) -> Option<Rc<RefCell<TreeNode>>> {
        Some(Rc::new(RefCell::new(TreeNode {val, left, right})))
    }

    #[test]
    fn test_1() {
        assert_eq!(
            Solution::build_tree(
                vec![3,9,20,15,7], vec![9,3,15,20,7]
            ), 
            node(3, node(9, None, None), node(20, node(15, None, None), node(7, None, None)))
        );
    }    

    #[test]
    fn test_2() {
        assert_eq!(
            Solution::build_tree(vec![-1], vec![-1]), 
            node(-1, None, None)
        );
    }    
}
